src/wasm-interpreter.h

```diff
@@ -454,28 +454,31 @@ class ModuleInstance {
         assert(left.type == curr->left->type);
         assert(right.type == curr->right->type);
         if (left.type == i32) {
-          int32_t l = left.geti32(), r = right.geti32();
+          uint32_t l = left.geti32(), r = right.geti32();
+          int32_t l_signed = l, r_signed = r;
           switch (curr->op) {
             case Add:      return Literal(l + r);
             case Sub:      return Literal(l - r);
             case Mul:      return Literal(l * r);
             case DivS: {
-              if (r == 0) trap("i32.div_s by 0");
-              if (l == INT32_MIN && r == -1) trap("i32.div_s overflow"); // signed division overflow
-              return Literal(l / r);
+              if (r_signed == 0) trap("i32.div_s by 0");
+              if (l_signed == INT32_MIN && r_signed == -1)
+                trap("i32.div_s overflow"); // signed division overflow
+              return Literal(l_signed / r_signed);
             }
             case DivU: {
               if (r == 0) trap("i32.div_u by 0");
-              return Literal(int32_t(uint32_t(l) / uint32_t(r)));
+              return Literal(uint32_t(l / r));
             }
             case RemS: {
-              if (r == 0) trap("i32.rem_s by 0");
-              if (l == INT32_MIN && r == -1) return Literal(int32_t(0));
-              return Literal(l % r);
+              if (r_signed == 0) trap("i32.rem_s by 0");
+              if (l_signed == INT32_MIN && r_signed == -1)
+                return Literal(int32_t(0));
+              return Literal(l_signed % r_signed);
```

Comment:
I guess they are all unneccessary, and pretty clear now that the types are unambiguous and they all match. Removed.